app/decorators/visit_decorator.rb

```diff
@@ -26,7 +29,9 @@ def slots
   end
 
   def rejection
-    @rejection ||= (object.rejection || object.build_rejection).decorate
+    @rejection ||= (object.rejection || object.build_rejection).decorate.tap do |rej|
+      rej.apply_nomis_reasons(self)
```

Comment:
On the second point. `VisitDecorator.prisoner_details_incorrect?` is explicitly called from `RejectionDecorator` - since it doesn't make sense to move this method to `NomisChecker` probably best leave the delegations as they are and continue to pass `self`.